spring-boot-project/spring-boot-tools/spring-boot-configuration-processor/src/main/java/org/springframework/boot/configurationprocessor/metadata/JsonConverter.java

```diff
@@ -160,26 +160,18 @@ private Object extractItemValue(Object value) {
 
 	private static class ItemMetadataComparator implements Comparator<ItemMetadata> {
 
+		private final Comparator<ItemMetadata> itemComparator = Comparator.comparing(this::isDeprecated)
+				.thenComparing(ItemMetadata::getName).thenComparing(ItemMetadata::getSourceType);
+
+		private final Comparator<ItemMetadata> groupComparator = Comparator.comparing(ItemMetadata::getName)
+				.thenComparing(ItemMetadata::getSourceType);
+
 		@Override
 		public int compare(ItemMetadata o1, ItemMetadata o2) {
 			if (o1.isOfItemType(ItemType.GROUP)) {
-				return compareGroup(o1, o2);
-			}
-			return compareProperty(o1, o2);
-		}
-
-		private int compareGroup(ItemMetadata o1, ItemMetadata o2) {
-			return o1.getName().compareTo(o2.getName());
-		}
-
-		private int compareProperty(ItemMetadata o1, ItemMetadata o2) {
-			if (isDeprecated(o1) && !isDeprecated(o2)) {
-				return 1;
-			}
-			if (isDeprecated(o2) && !isDeprecated(o1)) {
-				return -1;
+				return this.groupComparator.compare(o1, o2);
 			}
-			return o1.getName().compareTo(o2.getName());
+			return this.itemComparator.compare(o1, o2);
```

Comment:
If `o1` is of `GROUP` type and `o2` is of `PROPERTY` type, comparison is made except for `this::isDeprecated`, but `o1` is of `PROPERTY` type and `o2` is of `GROUP` type, it includes `this::isDeprecated` to compare. Isn't it inconsistent?

Wouldn't it be better to always compare `this::isDeprecated` regardless of type?

Reply:
This is not possible. Sort is applied after filtering of items to have just one type: https://github.com/spring-projects/spring-boot/blob/400b98700db7e5075888cba7bc50277b0d9a085b/spring-boot-project/spring-boot-tools/spring-boot-configuration-processor/src/main/java/org/springframework/boot/configurationprocessor/metadata/JsonConverter.java#L42-L43